scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -323,11 +344,184 @@ def highly_variable_genes(
         adata.var['highly_variable'] = df['highly_variable'].values
         adata.var['means'] = df['means'].values
         adata.var['dispersions'] = df['dispersions'].values
-        adata.var['dispersions_norm'] = df['dispersions_norm'].values.astype('float32', copy=False)
+        adata.var['dispersions_norm'] = df['dispersions_norm'].values.astype(
+            'float32', copy=False
+        )
+        if batch_key is not None:
+            adata.var['highly_variable_nbatches'] = df[
+                'highly_variable_nbatches'
+            ].values
+            adata.var['highly_variable_intersection'] = df[
+                'highly_variable_intersection'
+            ].values
+        if subset:
+            adata._inplace_subset_var(df['highly_variable'].values)
+    else:
+        return df
+
+
+def highly_variable_genes_norm_variance(
```

Comment:
I wanted to implement it in the same function, but wasn't sure how to:

1. deal with the fact that this requires count input and the others want log normalized (is there anything to check this?)
2. the main function is really hardcoded with the "dispersion" type keys and it seemed like adding this function would make the docstring really complicated.

any thoughts on this @gokceneraslan ?

Reply:
> I wanted to implement it in the same function, but wasn't sure how to:
> 
> 1. deal with the fact that this requires count input and the others want log normalized (is there anything to check this?)

We store log base under `.uns`, see: 

https://github.com/theislab/scanpy/blob/master/scanpy/preprocessing/_simple.py#L353

but it's not super reliable because there might be older anndata files that are generated before we introduced that feature.

Actually, we already exponentiate the data in the seurat flavor of the HVG function, so it also expects data in the non-logarithmized scale, see:

https://github.com/theislab/scanpy/blob/master/scanpy/preprocessing/_highly_variable_genes.py#L45

I think the reason we require log-transformed data is that cellranger flavor uses `mean(log(X))` but the seurat flavor uses `log(mean(X))`, so if flavor is seurat we technically do `log(mean(exp(log(X))))` 😄 

> 2. the main function is really hardcoded with the "dispersion" type keys and it seemed like adding this function would make the docstring really complicated.

I see, yes it'd be a bit ugly. But I think if we write that min/max_dispersion and min/max_mean parameters are ignored if flavor is seurat_v3, it'd be ok. We already have n_top_genes parameter that the new flavor can require.

This overall requires big changes though, @ivirshup wdyt?

Reply:
> We store log base under .uns, see:

> 
> https://github.com/theislab/scanpy/blob/master/scanpy/preprocessing/_simple.py#L353
> 
> but it's not super reliable because there might be older anndata files that are generated before we introduced that feature.

But for absolute consistency with seurat v3, we would have to check that user has inputted non-negative integers (i.e., no library size normalization). Is there a method for this? It seems like it could be a useful util down the road as already some external methods require this as well.

Reply:
Hmm I see. There is no nice way to do this I think, but we can check if values are whole numbers via 

```python
np.all(np.mod(adata.X.data[:100], 1) == 0)
```

or we can also check if data is TPM normalized:

```python
sums = adata.X.sum(1)
np.all(sums == sums[0])
```

Also we can make `sc.pp.normalize_total` function add something to adata.uns like 

```python
adata.uns['normalize_total'] = {}
adata.uns['normalize_total']['totals'] = adata.X.sum(1)
adata.uns['normalize_total']['target_sum'] = 10000
```

This way we can also add an `sc.pp.unnormalize` function. It would be nice to have sc.pp.expm1() and sc.pp.unnormalize() in general :)

Would these help?

Reply:
(There is also https://github.com/theislab/scanpy/blob/master/scanpy/external/pp/_scvi.py#L118)

Reply:
I can do this, just want to confirm we have consensus on these changes? This will indeed be a big change.

Reply:
I think it'd be useful to have an internal "_is_raw()" function in utils.py or so to check if adata.X has counts. This can then be utilized in many other functions like my sctransform PR, or DCA or scVI etc.

Reply:
I added `check_nonnegative_integers()` util and decided for now to add a `layer` option, where a user can select a layer to use for computing HVGs. As computing HVGs with `seurat_v3` flavor would be one of the first steps after basic gene and cell filtering, it seems that this is a suitable compromise for now. An issue with what you propose for storing the size factors is that `normalize_total` currently has an option to store the size factor for you and it's off by default. So I can still do this and add `sc.pp.unnormalize` but it seems like it would involve changing the default of that function. Thoughts?

Reply:
`check_nonnegative_integers()` looks great, can you add a test too if it's not too much to ask? Layer option is also great I think, because many people store counts in `.layer['counts']`.

I didn't know about the key_added option of normalize_total, interesting. Then sc.pp.normalize can require a mandatory "key" option which is an obs key where size factors are stored. So it can only be used if the user used the key_added option of normalize_total. 

Btw, I think sc.pp.unnormalize idea is beyond the scope of this PR so it's completely up to you. I was just thinking out loud about all options regarding raw counts :) But if you feel like implementing it then a separate PR would be a better place.